Library/Homebrew/dependency.rb

```diff
@@ -124,11 +124,39 @@ def merge_repeats(all)
       all.map(&:name).uniq.map do |name|
         deps = grouped.fetch(name)
         dep  = deps.first
-        tags = deps.flat_map(&:tags).uniq
+        tags = merge_tags(deps)
         option_names = deps.flat_map(&:option_names).uniq
         dep.class.new(name, tags, dep.env_proc, option_names)
       end
     end
+
+    private
+
+    def merge_tags(deps)
+      options = deps.flat_map(&:option_tags).uniq
+      merge_necessity(deps) + merge_temporality(deps) + options
+    end
+
+    def merge_necessity(deps)
+      if deps.any? { |dep| !dep.recommended? && !dep.optional? }
+        # FIXME: Should be `required?`, but `required?` implies `!build?`.
+        [] # Implies required dependency.
+      elsif deps.any?(&:recommended?)
+        [:recommended]
+      else # deps.all?(&:optional?)
+        [:optional]
+      end
```

Comment:
I wonder whether it would help to make the code more readable by implementing this function in this way:
```ruby
def merge_necessity(deps)
  recommended = false
  deps.each do |dep|
    if dep.recommended?
      recommended = true
    elsif !dep.optional?
     return []
    end
  end
  recommended ? [:recommended] : [:optional]
end
```

As a bonus, it will only enumerate the list for once.